File: lodes/Sorting/Heapsort/heapsort.py
def heapify(values):
    """ 
        Note this process could be done inplace by replacing infinity in the 0th index
        with the last grabbed element. As an element is taken from values, another space
        is made available meaning an element could be entered into 1, 2 ...
    """
    # Heap starts at index 1
    heap = [float("inf")]

    # Add all values while maintaining sort/shape property
    for i in range(len(values)):
        heap.append(values[i])
        
        # Check parent is greater and if not swap
        j = i + 1
        while heap[j//2] < heap[j]:
            tmp = heap[j//2]
            heap[j//2] = heap[j]
            heap[j] = tmp

            j = j // 2 

    return heap[1:]

def heapsort(values):
    """ Takes in highest value of heap and re-heapifies """
    heap = []
    while len(values):
        tmp = heapify(values)
        heap.append(tmp[0])
        values = tmp[1:] 

    return reversed(heap)

File: lodes/Sorting/Heapsort/heapsort.c
#ifndef HEAPSORT_C
#define HEAPSORT_C

void heapify(int* values, int size, int index)
{
    int l = (index << 1) + 1;
    int r = l + 1;

    int max = index;
    if(l < size && values[l] > values[max]) max = l;
    if(r < size && values[r] > values[max]) max = r;

    if(max != index) {
        int tmp = values[index];
        values[index] = values[max];
        values[max] = tmp;

        heapify(values, size, max);
    }
}

void heapsort(int* values, int size)
{
    // Heapify
    for(int i = size / 2 - 1; i >= 0; i--) {
        heapify(values, size, i);
    }

    // Sort
    int len = size;
    while(len--) {
        int tmp = values[0];
        values[0] = values[len];
        values[len] = tmp;

        heapify(values, len, 0);
    }
}

#endif
